File: section_02/code4_7_FibonacciNumber.rb
#!/usr/bin/env ruby
# -*- coding: utf-8 -*-

require '../common/ruby/assert.rb'

#
# 再帰だと計算量が爆発するので、forで行うパターン
#

def fibo(n)
  return 0 if n == 0
  return 1 if n == 1

  a = 0
  b = 1
  (n - 1).times do
    tmp = (a + b)
    a = b
    b = tmp
  end

  b
end

Common::Assert.equal(0, fibo(0))
Common::Assert.equal(1, fibo(1))
Common::Assert.equal(1, fibo(2))
Common::Assert.equal(2, fibo(3))
Common::Assert.equal(3, fibo(4))
Common::Assert.equal(5, fibo(5))
Common::Assert.equal(75025, fibo(25))
Common::Assert.equal(12586269025, fibo(50))

pp "ok"

File: section_04/code4_7_FibonacciNumber.swift
import Foundation

// 再帰を使うと計算量が爆発するので
// for分の反復で求める

func fibo(_ n: Int) -> Int {
    if n == 0 { return 0; }
    if n == 1 { return 1; }
    var a = 0
    var b = 1
    for _ in 0..<(n - 1) {
        let tmp = (a + b)
        a = b
        b = tmp
    }

    return b
}

assert(fibo(0) == 0)
assert(fibo(1) == 1)
assert(fibo(2) == 1)
assert(fibo(3) == 2)
assert(fibo(4) == 3)
assert(fibo(5) == 5)

print("25 = \(fibo(25))\n") 
print("50 = \(fibo(50))\n") // 12586269025


